moveit_core/online_signal_smoothing/src/acceleration_filter.cpp

```diff
@@ -0,0 +1,299 @@
+    x.assign(M.nonZeros(), 0.0);
+    sparse_matrix.x = x.data();
+
+    update(M);
+  }
+
+  void update(Eigen::SparseMatrix<double>& M)
+  {
+    for (size_t ind = 0; ind < i.size(); ++ind)
+    {
+      i[ind] = M.innerIndexPtr()[ind];
+    }
+
+    for (size_t ind = 0; ind < j.size(); ++ind)
+    {
+      j[ind] = M.outerIndexPtr()[ind];
+    }
+    for (size_t ind = 0; ind < x.size(); ++ind)
+    {
+      x[ind] = M.data().at(ind);
+    }
+  }
+};
+
+MOVEIT_STRUCT_FORWARD(OSQPDataWrapper);
+
+struct OSQPDataWrapper
+{
+  OSQPDataWrapper(Eigen::SparseMatrix<double>& P_sparse, Eigen::SparseMatrix<double>& A_sparse)
+    : P{ P_sparse }, A{ A_sparse }
+  {
+    data.n = P_sparse.rows();
+    data.m = A_sparse.rows();
+    data.P = &P.sparse_matrix;
+    q = Eigen::VectorXd::Zero(P_sparse.rows());
+    data.q = q.data();
+    data.A = &A.sparse_matrix;
+    l = Eigen::VectorXd::Zero(A_sparse.rows());
+    data.l = l.data();
+    u = Eigen::VectorXd::Zero(A_sparse.rows());
+    data.u = u.data();
+  }
+
+  void update_A(OSQPWorkspace* work, Eigen::SparseMatrix<double>& A_sparse)
+  {
+    A_sparse.makeCompressed();
+    A.update(A_sparse);
+    osqp_update_A(work, A.x.data(), OSQP_NULL, A.x.size());
+  }
+
+  CSCWrapper P;
+  CSCWrapper A;
+  Eigen::VectorXd q;
+  Eigen::VectorXd l;
+  Eigen::VectorXd u;
+  OSQPData data{};
+};
+bool AccelerationLimitedPlugin::initialize(rclcpp::Node::SharedPtr node, moveit::core::RobotModelConstPtr robot_model,
+                                           size_t num_joints)
+{
+  // copy inputs into member variables
+  node_ = node;
+  num_joints_ = num_joints;
+  robot_model_ = robot_model;
+  cur_acceleration_ = Eigen::VectorXd::Zero(num_joints);
+
+  // get node parameters and store in member variables
+  auto param_listener = online_signal_smoothing::ParamListener(node_);
+  auto params = param_listener.get_params();
+  update_rate_ = params.update_rate;
+  move_group_name_ = params.move_group_name;
+
+  // get robot acceleration limits and store in member variables
+  auto joint_model_group = robot_model_->getJointModelGroup(move_group_name_);
+  auto joint_bounds = joint_model_group->getActiveJointModelsBounds();
+  min_acceleration_limits_ = Eigen::VectorXd::Zero(num_joints);
+  max_acceleration_limits_ = Eigen::VectorXd::Zero(num_joints);
+  size_t ind = 0;
+  for (const auto& joint_bound : joint_bounds)
+  {
+    for (const auto& variable_bound : *joint_bound)
+    {
+      if (variable_bound.acceleration_bounded_)
+      {
+        min_acceleration_limits_[ind] = variable_bound.min_acceleration_;
+        max_acceleration_limits_[ind] = variable_bound.max_acceleration_;
+      }
+      else
+      {
+        RCLCPP_ERROR(node_->get_logger(), "The robot must have acceleration joint limits specified for all joint to "
+                                          "use AccelerationLimitedPlugin.");
+        return false;
+      }
+    }
+    ind++;
+  }
+
+  // setup osqp optimization problem
```

Comment:
Do you mind adding more documentation here? I assume that this code might be a reference for other applications in moveit that could use osqp so more info will be very helpful in the future

Reply:
I added the math formulation